File: python/exer07a_data_server.py
'''
THE DATA SERVER PROBLEM
Version A: Solving the problem using a condition variable
'''

from dataclasses import dataclass
import time
import threading



@dataclass
class Counter:
    value: int = 0
    cond: threading.Condition = threading.Condition()



def check_auth_user():
    print('[   Auth   ] Start')
    # Send request to authenticator, check permissions, encrypt, decrypt...
    time.sleep(20)
    print('[   Auth   ] Done')



def process_files(lst_file_name: list, counter: Counter):
    for file_name in lst_file_name:
        # Read file
        print('[ ReadFile ] Start', file_name)
        time.sleep(10)
        print('[ ReadFile ] Done ', file_name)

        with counter.cond:
            counter.value -= 1
            counter.cond.notify()

        # Write log into disk
        time.sleep(5)
        print('[ WriteLog ]')



def process_request():
    lst_file_name = [ 'foo.html', 'bar.json' ]
    counter = Counter(value=len(lst_file_name))

    # The server checks auth user while reading files, concurrently
    threading.Thread(target=process_files, args=(lst_file_name, counter)).start()
    check_auth_user()

    # The server waits for completion of loading files
    with counter.cond:
        while counter.value > 0:
            counter.cond.wait(10) # timeout = 10 seconds

    print('\nNow user is authorized and files are loaded')
    print('Do other tasks...\n')



process_request()

File: csharp/exercise/exer07a-data-server.cs
/*
 * THE DATA SERVER PROBLEM
 * Version A: Solving the problem using a condition variable
 */
using System;
using System.Threading;



class Exer07A : IRunnable
{
    public void run()
    {
        var server = new DataServer();
        server.processRequest();
    }


    private class DataServer
    {
        private class Counter
        {
            public int value;
            public Counter(int value)
            {
                this.value = value;
            }
        }


        public void processRequest()
        {
            var lstFileName = new String[] { "foo.html", "bar.json" };
            var counter = new Counter(lstFileName.Length);

            // The server checks auth user while reading files, concurrently
            new Thread(() => processFiles(lstFileName, counter)).Start();
            checkAuthUser();

            // The server waits for completion of loading files
            lock (counter)
            {
                while (counter.value > 0)
                {
                    Monitor.Wait(counter, 10000); // timeout = 10 seconds
                }
            }

            Console.WriteLine("\nNow user is authorized and files are loaded");
            Console.WriteLine("Do other tasks...\n");
        }


        // This task consumes CPU (and network bandwidth, maybe)
        private void checkAuthUser()
        {
            Console.WriteLine("[   Auth   ] Start");
            // Send request to authenticator, check permissions, encrypt, decrypt...
            Thread.Sleep(20000);
            Console.WriteLine("[   Auth   ] Done");
        }


        // This task consumes disk
        private void processFiles(String[] lstFileName, Counter counter)
        {
            foreach (var fileName in lstFileName)
            {
                // Read file
                Console.WriteLine("[ ReadFile ] Start " + fileName);
                Thread.Sleep(10000);
                Console.WriteLine("[ ReadFile ] Done  " + fileName);

                lock (counter)
                {
                    --counter.value;
                    Monitor.Pulse(counter);
                }

                // Write log into disk
                Thread.Sleep(5000);
                Console.WriteLine("[ WriteLog ]");
            }
        }
    }
}
